### DISCUSSION TITLE
path ValidationError

### DISCUSSION THREAD

**Discussion by rsaleev** [fastapi/fastapi#10757] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.get(
    "/{id}/billing",
    response_model=List[OrderBillingView],
)
async def get_order_billing(
    session: Annotated[AsyncSession, Depends(connection)], id: int
):
    stmt = get_order_billing_data_by_id(id)
    result = await session.execute(stmt)
    return [OrderBillingView.model_validate(item) for item in result.scalars()]


@router.get(
    "/states",
    response_model=OrderStatesView,

)
async def list_states(request):
    return OrderStatesView.model_validate(
        [item.name for item in OrderStatePydanticEnum]
    )


@router.get(
    "/orgs",
    response_model=OrgsView,

)
async def list_orgs(request):
    return OrgsView.model_validate([item.name for item in OwnerEnum])
```

### Description

```
 raise RequestValidationError(_normalize_errors(errors), body=body)
fastapi.exceptions.RequestValidationError: [{'type': 'int_parsing', 'loc': ('path', 'id'), 'msg': 'Input should be a valid integer, unable to parse string as an integer', 'input': 'orgs', 'url': 'https://errors.pydantic.dev/2.5/v/int_parsing'}]
```
But if put these endpoints before I use path param everything works fine

### Operating System

Linux

### Operating System Details

Fedora 39

### FastAPI Version

0.104.1

### Pydantic Version

2.5.2

### Python Version

3.12

**Comment by JavierSanchezCastro** [comment]
Please provide a full basic example to test and reproduce your error

**Comment by kiraware** [comment]
The order of path in fastapi is very important. You have path in order like this
- `/{id}/billing` <- fastapi use this path
- `/states`
- `/orgs`

Based on your error message i guess you trying to access `/orgs`. But fastapi interpret the `/orgs` with this path `/{id}/billing` because this is dynamic path and has higher order. `{id}` accept integer, while the input is `"orgs"`. Hence it complaint parsing string to integer. So the solution is change the order to
- `/states`
- `/orgs` <- fastapi use this path
- `/{id}/billing`

So that when you access `/orgs` it will success because fastapi found this path first. Hope it answer your question.

**Comment by YuriiMotov** [comment]
Didn't manage to reproduce this.
It was probably an issue in specific version of Starlette.

```python
from fastapi import FastAPI, APIRouter
from fastapi.testclient import TestClient


app = FastAPI()


router = APIRouter()


@router.get(
    "/{id}/billing",
)
async def get_order_billing(id: int):
    return "get_order_billing"


@router.get("/states")
async def list_states():
    return "list_states"


@router.get("/orgs")
async def list_orgs():
    return "list_orgs"


app.include_router(router=router)


def test_():
    client = TestClient(app)

    resp = client.get("/orgs")
    assert resp.status_code == 200
    assert resp.json() == "list_orgs"
```
